go-controller/pkg/ovn/logical_switch_manager.go

```diff
@@ -0,0 +1,255 @@
+		if err != nil {
+			klog.Errorf("IPAM for subnet %s was not initialized for node %q", subnet, nodeName)
+			return err
+		}
+		ipams = append(ipams, ipam)
+	}
+	manager.cache[nodeName] = logicalSwitchInfo{
+		hostSubnets:  hostSubnets,
+		ipams:        ipams,
+		noHostSubnet: len(hostSubnets) == 0,
+	}
+
+	return nil
+}
+
+// AddNoHostSubnetNode adds/updates a node without any host subnets
+// to the logical switch manager
+func (manager *logicalSwitchManager) AddNoHostSubnetNode(nodeName string) error {
+	// setting the hostSubnets slice argument to nil in the cache means an object
+	// exists for the switch but it was not assigned a hostSubnet by ovn-kube
+	// this will be true for nodes that are marked as host-subnet only.
+	return manager.AddNode(nodeName, nil)
+}
+
+// Remove a switch/node from the the logical switch manager
+func (manager *logicalSwitchManager) DeleteNode(nodeName string) {
+	manager.Lock()
+	defer manager.Unlock()
+	delete(manager.cache, nodeName)
+}
+
+// Given a switch name, checks if the switch is a noHostSubnet switch
+func (manager *logicalSwitchManager) IsNonHostSubnetSwitch(nodeName string) bool {
+	manager.RLock()
+	defer manager.RUnlock()
+	lsi, ok := manager.cache[nodeName]
+	return ok && lsi.noHostSubnet
+}
+
+// Given a switch name, get all its host-subnets
+func (manager *logicalSwitchManager) GetSwitchSubnets(nodeName string) []*net.IPNet {
+	manager.RLock()
+	defer manager.RUnlock()
+	lsi, ok := manager.cache[nodeName]
+	// make a deep-copy of the underlying slice and return so that there is no
+	// resource contention
+	if ok && len(lsi.hostSubnets) > 0 {
+		subnets := make([]*net.IPNet, len(lsi.hostSubnets))
+		for i, hsn := range lsi.hostSubnets {
+			subnet := *hsn
+			subnets[i] = &subnet
+		}
+		return subnets
+	}
+	return nil
+}
+
+// AllocateIPs will block off IPs in the ipnets slice as already allocated
+// for a given switch
+func (manager *logicalSwitchManager) AllocateIPs(nodeName string, ipnets []*net.IPNet) error {
+	manager.RLock()
+	defer manager.RUnlock()
+	lsi, ok := manager.cache[nodeName]
+	if len(ipnets) <= 0 || !ok || (ok && len(lsi.ipams) <= 0) {
+		return fmt.Errorf("unable to allocate ips %v for node: %s",
+			ipnets, nodeName)
+
+	}
+	for _, ipnet := range ipnets {
+		for _, ipam := range lsi.ipams {
+			cidr := ipam.CIDR()
+			if cidr.Contains(ipnet.IP) {
+				if err := ipam.Allocate(ipnet.IP); err != nil {
+					return err
+				}
+				break
+			}
+		}
+	}
+	return nil
+}
+
+// AllocateNextIPs allocates IP addresses from each of the host subnets
+// for a given switch
+func (manager *logicalSwitchManager) AllocateNextIPs(nodeName string) ([]*net.IPNet, error) {
+	manager.RLock()
+	defer manager.RUnlock()
+	var ipnets []*net.IPNet
+	var ip net.IP
+	var err error
+	lsi, ok := manager.cache[nodeName]
+	if ok && len(lsi.ipams) == 0 {
+		return nil, fmt.Errorf("failed to allocate IPs for node %s because there is no IPAM instance", nodeName)
+	}
```

Comment:
what if `ok` if false? That is while you were waiting on the `RLock()` at L187, the `lsi` disappeared on you? Should we check return error if `!ok`?

Reply:
done.